cd test-project && cargo +nightly hack test --feature-powerset fails with duplicate rlibs

Running the tests twice (with different features) results in pollution:

```
/src/compiletest-rs/test-project$ find target/ -name 'libtestp*.rlib' -type f | xargs ls -lah
-rw-rw-r-- 1 tamird tamird 3.3K Jun  2 12:02 target/debug/deps/libtestp-4450211ae8c20f2d.rlib
-rw-rw-r-- 1 tamird tamird 3.3K Jun  2 12:13 target/debug/deps/libtestp-c6f9abab7ee2589d.rlib
```

which leads to failure:

```
running 5 tests
test [run-pass] run-pass/with-own-crate.rs ... FAILED
test [run-pass] run-pass/associated-types-binding-in-where-clause.rs ... ok
test [run-pass] run-pass/issue-10025.rs ... ok
test [run-pass] run-pass/while-with-break.rs ... ok
test [run-pass] run-pass/vec-slice-drop.rs ... ok

failures:

failures:
    [run-pass] run-pass/with-own-crate.rs

test result: FAILED. 4 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.09s

test compile_test ... FAILED

failures:

---- compile_test stdout ----

error: compilation failed!
status: exit status: 1
command: "rustc" "tests/run-pass/with-own-crate.rs" "-L" "/tmp/compiletestLDU13B" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestLDU13B/with-own-crate.stage-id" "-L" "/home/tamird/src/compiletest-rs/test-project/target/debug/deps" "-L" "/home/tamird/src/compiletest-rs/test-project/target/debug" "-L" "/home/tamird/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/tamird/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib" "-L" "/tmp/compiletestLDU13B/with-own-crate.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"multiple candidates for `rlib` dependency `testp` found","code":{"code":"E0464","explanation":"The compiler found multiple library files with the requested crate name.\n\n```compile_fail\n// aux-build:crateresolve-1.rs\n// aux-build:crateresolve-2.rs\n// aux-build:crateresolve-3.rs\n\nextern crate crateresolve;\n//~^ ERROR multiple candidates for `rlib` dependency `crateresolve` found\n\nfn main() {}\n```\n\nThis error can occur in several different cases -- for example, when using\n`extern crate` or passing `--extern` options without crate paths. It can also be\ncaused by caching issues with the build directory, in which case `cargo clean`\nmay help.\n\nIn the above example, there are three different library files, all of which\ndefine the same crate name. Without providing a full path, there is no way for\nthe compiler to know which crate it should use.\n"},"level":"error","spans":[{"file_name":"tests/run-pass/with-own-crate.rs","byte_start":0,"byte_end":19,"line_start":1,"line_end":1,"column_start":1,"column_end":20,"is_primary":true,"text":[{"text":"extern crate testp;","highlight_start":1,"highlight_end":20}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"candidate #1: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-4450211ae8c20f2d.rlib","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"candidate #2: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-c6f9abab7ee2589d.rlib","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0464]: multiple candidates for `rlib` dependency `testp` found\n --> tests/run-pass/with-own-crate.rs:1:1\n  |\n1 | extern crate testp;\n  | ^^^^^^^^^^^^^^^^^^^\n  |\n  = note: candidate #1: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-4450211ae8c20f2d.rlib\n  = note: candidate #2: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-c6f9abab7ee2589d.rlib\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
{"message":"For more information about this error, try `rustc --explain E0464`.","code":null,"level":"failure-note","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0464`.\n"}

------------------------------------------

thread 'compile_test' panicked at 'Some tests failed', /home/tamird/src/compiletest-rs/src/lib.rs:113:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    compile_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.18s

```

Prettified JSON:
```json
{
  "message": "multiple candidates for `rlib` dependency `testp` found",
  "code": {
    "code": "E0464",
    "explanation": "The compiler found multiple library files with the requested crate name.\n\n```compile_fail\n// aux-build:crateresolve-1.rs\n// aux-build:crateresolve-2.rs\n// aux-build:crateresolve-3.rs\n\nextern crate crateresolve;\n//~^ ERROR multiple candidates for `rlib` dependency `crateresolve` found\n\nfn main() {}\n```\n\nThis error can occur in several different cases -- for example, when using\n`extern crate` or passing `--extern` options without crate paths. It can also be\ncaused by caching issues with the build directory, in which case `cargo clean`\nmay help.\n\nIn the above example, there are three different library files, all of which\ndefine the same crate name. Without providing a full path, there is no way for\nthe compiler to know which crate it should use.\n"
  },
  "level": "error",
  "spans": [
    {
      "file_name": "tests/run-pass/with-own-crate.rs",
      "byte_start": 0,
      "byte_end": 19,
      "line_start": 1,
      "line_end": 1,
      "column_start": 1,
      "column_end": 20,
      "is_primary": true,
      "text": [
        {
          "text": "extern crate testp;",
          "highlight_start": 1,
          "highlight_end": 20
        }
      ],
      "label": null,
      "suggested_replacement": null,
      "suggestion_applicability": null,
      "expansion": null
    }
  ],
  "children": [
    {
      "message": "candidate #1: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-4450211ae8c20f2d.rlib",
      "code": null,
      "level": "note",
      "spans": [],
      "children": [],
      "rendered": null
    },
    {
      "message": "candidate #2: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-c6f9abab7ee2589d.rlib",
      "code": null,
      "level": "note",
      "spans": [],
      "children": [],
      "rendered": null
    }
  ],
  "rendered": "error[E0464]: multiple candidates for `rlib` dependency `testp` found\n --> tests/run-pass/with-own-crate.rs:1:1\n  |\n1 | extern crate testp;\n  | ^^^^^^^^^^^^^^^^^^^\n  |\n  = note: candidate #1: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-4450211ae8c20f2d.rlib\n  = note: candidate #2: /home/tamird/src/compiletest-rs/test-project/target/debug/deps/libtestp-c6f9abab7ee2589d.rlib\n\n"
}
{
  "message": "aborting due to previous error",
  "code": null,
  "level": "error",
  "spans": [],
  "children": [],
  "rendered": "error: aborting due to previous error\n\n"
}
{
  "message": "For more information about this error, try `rustc --explain E0464`.",
  "code": null,
  "level": "failure-note",
  "spans": [],
  "children": [],
  "rendered": "For more information about this error, try `rustc --explain E0464`.\n"
}

```

Perhaps we should have a `Config::clean_rlib()`, akin to [`Config::clean_rmeta()`](https://docs.rs/compiletest_rs/0.10.2/compiletest_rs/common/struct.Config.html#method.clean_rmeta) for dealing with this problem?

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -24,11 +24,9 @@ jobs:
     - uses: actions/checkout@v3
 
     - name: Install rust toolchain
-      uses: actions-rs/toolchain@v1
+      uses: dtolnay/rust-toolchain@master
       with:
-        profile: minimal
         toolchain: ${{ matrix.rust }}
-        override: true
 
     - name: Run tests
       run: cargo test
@@ -48,11 +46,8 @@ jobs:
     - uses: actions/checkout@v3
 
     - name: Install rust toolchain
-      uses: actions-rs/toolchain@v1
+      uses: dtolnay/rust-toolchain@nightly
       with:
-        profile: minimal
-        toolchain: nightly
-        override: true
         components: rustc-dev, llvm-tools
 
     - name: Run tests

--- a/README.md
+++ b/README.md
@@ -86,8 +86,8 @@ let mut config = compiletest::Config::default();
 config.link_deps();
 ```
 
-Note that `link_deps()` should not be used if any of the added paths contain
-spaces, as these are currently not handled correctly.
+Note that `link_deps()` panics if any of the added paths contain spaces, as
+these are currently not handled correctly.
 
 Example
 -------

--- a/src/common.rs
+++ b/src/common.rs
@@ -286,8 +286,10 @@ impl Config {
         // Append to current flags if any are set, otherwise make new String
         let mut flags = self.target_rustcflags.take().unwrap_or_default();
         for p in env::split_paths(&lib_paths) {
+            let p = p.to_str().unwrap();
+            assert!(!p.contains(' '), "spaces in paths not supported: {}", p);
             flags += " -L ";
-            flags += p.to_str().unwrap();
+            flags += p;
         }
 
         self.target_rustcflags = Some(flags);

--- a/src/common.rs
+++ b/src/common.rs
@@ -281,18 +281,13 @@ impl Config {
 
         // Dependencies can be found in the environment variable. Throw everything there into the
         // link flags
-        let lib_paths = env::var(varname).unwrap_or_else(|err| match err {
-            env::VarError::NotPresent => String::new(),
-            err => panic!("can't get {} environment variable: {}", varname, err),
-        });
+        let lib_paths = env::var_os(varname).unwrap_or_default();
 
         // Append to current flags if any are set, otherwise make new String
-        let mut flags = self.target_rustcflags.take().unwrap_or_else(String::new);
-        if !lib_paths.is_empty() {
-            for p in env::split_paths(&lib_paths) {
-                flags += " -L ";
-                flags += p.to_str().unwrap(); // Can't fail. We already know this is unicode
-            }
+        let mut flags = self.target_rustcflags.take().unwrap_or_default();
+        for p in env::split_paths(&lib_paths) {
+            flags += " -L ";
+            flags += p.to_str().unwrap();
         }
 
         self.target_rustcflags = Some(flags);

--- a/src/common.rs
+++ b/src/common.rs
@@ -295,26 +295,61 @@ impl Config {
         self.target_rustcflags = Some(flags);
     }
 
+    fn find_deps_with_extension(&self, ext: &'static str) -> impl Iterator<Item = PathBuf> + '_ {
+        self.target_rustcflags
+            .iter()
+            .flat_map(|flags| flags.split_whitespace().filter(|s| s.ends_with("/deps")))
+            .filter_map(|directory| read_dir(directory).ok())
+            .flat_map(move |entries| {
+                entries.filter_map(Result::ok).filter_map(move |entry| {
+                    let path = entry.path();
+                    let extension = path.extension()?;
+                    (extension == ext).then_some(path)
+                })
+            })
+    }
+
     /// Remove rmeta files from target `deps` directory
     ///
     /// These files are created by `cargo check`, and conflict with
     /// `cargo build` rlib files, causing E0464 for tests which use
     /// the parent crate.
     pub fn clean_rmeta(&self) {
-        if self.target_rustcflags.is_some() {
-            for directory in self.target_rustcflags
-                .as_ref()
-                .unwrap()
-                .split_whitespace()
-                .filter(|s| s.ends_with("/deps"))
-            {
-                if let Ok(mut entries) = read_dir(directory) {
-                    while let Some(Ok(entry)) = entries.next() {
-                        if entry.file_name().to_string_lossy().ends_with(".rmeta") {
-                            let _ = remove_file(entry.path());
+        self.find_deps_with_extension("rmeta").for_each(|path| {
+            let _: Result<(), _> = remove_file(path);
+        })
+    }
+
+    /// Remove "duplicate" rlib files from target `deps` directory
+    ///
+    /// These files are created by `cargo build`; running the tests with
+    /// multiple sets of features can produce multiple rlib files for
+    /// each crate, causing E0464 for tests which use the parent crate.
+    pub fn clean_rlib(&self) {
+        let mut rlibs = self.find_deps_with_extension("rlib").collect::<Vec<_>>();
+        let () = rlibs.sort();
+        let mut rlibs = rlibs.as_slice();
+        loop {
+            match rlibs {
+                [] => break,
+                [first, rest @ ..] => match rest {
+                    [] => break,
+                    [second, ..] => {
+                        fn stem(path: &PathBuf) -> &str {
+                            let stem = (|| {
+                                let file_name = path.file_name()?;
+                                let file_name = file_name.to_str()?;
+                                let (stem, _) = file_name.split_once('-')?;
+                                Some(stem)
+                            })();
+                            stem.unwrap_or_else(|| panic!("unable to parse {}", path.display()))
                         }
+                        if stem(first) == stem(second) {
+                            let _: Result<(), _> = remove_file(first);
+                        }
+                        rlibs = rest;
                     }
-                }
+                },
             }
         }
     }

--- a/test-project/tests/tests.rs
+++ b/test-project/tests/tests.rs
@@ -18,6 +18,7 @@ fn run_mode(mode: &'static str, custom_dir: Option<&'static str>) {
             .into(),
     );
     config.clean_rmeta();
+    config.clean_rlib();
     config.strict_headers = true;
 
     compiletest::run_tests(&config);

